rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+
+
+Also, these contracts, when provided, would also significantly improve Kani’s verification time since all loops would be unrolled only to a single iteration, as opposed to a number of iterations which is Kani’s current behavior. For example, although the following loop is bounded and we can easily find out the unwinding number `50000` for it, unwinding the loop 50000 times will lead to a large query, and hence slow down Kani. With loop contract, we only verify the single abstract iteration of the loop, which leads to a smaller query.
+```rust
+#[kani::proof]
+#[kani::unwind(50000)]
+fn main() {
+    let mut y: i32 = 50000;
+
+    while y > 0 {
+        y = y - 1;
+    }
+    assert!(y == 0);
+}
+```
+
+
+## User Experience
+
+Once this RFC has been stabilized, users will be able to synthesize loop contracts for loops and verify programs with the synthesized loop contracts with a new command line flag `--synthesize-loop-contracts`.
+
+### How the Synthesizer Works with Unwinding Numbers
+For each loop, users will be able to choose to unwind it for a certain number of times by specifying the unwinding numbers explicitly, or to synthesize loop contracts for it when no unwinding number specified. 
+
+We introduce a new way for users to specify unwinding numbers for a single loops rather than just for the whole functions: specifying loop numbers using labels. E.g., in the following program, both loops in the function `foo` will be unwound `5` times, while only the loop `inner` in the function `bar` will be unwound `2` times.
+```rust
+#[kani::unwind(5)]
+fn foo() {
+    let mut y: i32 = 50000;
+    let mut x: i32 = 2;
+    while y > 0 {
+        while x > 0 {
+            y = y - 1;
+            x = x - 1;
+        }
+        x = 2;
+    }
+    assert!(y == 0 || y == -1);
+}
+
+#[kani::unwind(inner:2)]
+fn bar() {
+    let mut y: i32 = 50000;
+    let mut x: i32 = 2;
+    `outer`: while y > 0 {
+        `inner`: while x > 0 {
+            y = y - 1;
+            x = x - 1;
+        }
+        x = 2;
+    }
+    assert!(y == 0 || y == -1);
+}
+```
+Then, we the flag `--synthesize-loop-contracts` is present, the synthesizer will synthesize loop contracts for only the loop `outer` in the function `bar` and unwind the other three loops.
+
+### Limit Resource Used by Synthesizer for Termination
+Without resource limit, an enumerative synthesizer may run forever to exhaust search space consist of infinite number of candidates, especially when there is no solution in the search space. So, for the guarantee of termination, we provide users options: <code>--limit-synthesis-time <ins>T</ins></code> to limit the running time of the synthesizer, and <code>--limit-synthesis-size <ins>N</ins></code> to limit the sizes (number of nodes in the AST of the candidate predicate) of loop-invariant candidates we search for during the enumeration.
```

Comment:
>    Does the synthesizer need to be applied to a whole program?

Yes, it will synthesize loop invariants for all loops at the same time.

>    Can we get something out of the synthesizer before it gets killed?

Yes, we plan to implement best-effort-synthesis.
 
>   Why not use memory instead of AST to refer to size?

We will use running time only.